File: src/aircraft/game/Canvas.java
package aircraft.game;

import javax.swing.JPanel;
import java.awt.image.BufferedImage;
import java.awt.Graphics;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.KeyAdapter;
import java.awt.Color;
import java.awt.Font;

import java.util.Set;
import java.util.HashSet;
import java.util.HashMap;
import java.util.Iterator;

import aircraft.game.plane.*;
import aircraft.game.bullet.*;
import aircraft.game.bomb.*;
import aircraft.game.supply.*;

public class Canvas extends JPanel {
  // Define serialVersionUID which is used during deserialization to verify
  // that the sender and receiver of a serialized object have loaded classes
  // for that object that are compatible with respect to serialization. 
  private static final long serialVersionUID = 1L;

  // The game status.

  private static final BufferedImage background = ImageLoader.readImg("background.png");
  private static final BufferedImage gameover = ImageLoader.readImg("gameover.png");

  // The fundamental components of graphics window.
  private final int width = Setting.width;
  private final int height = Setting.height;
  
  private final HeroPlane hero = // Initialize the hero plane (player).
    new HeroPlane(this, Setting.hero_init_pos.x, Setting.hero_init_pos.y);
  public boolean status = Setting.RUNNING;

  // The hash set to store all objects. Each time an object is constructed,
  // it should be added into this set, e.g. the push operation should be
  // called in the base constructor.
  public HashSet<Object> objects = new HashSet<>();
  public HashSet<Object> trash = new HashSet<>();
  public HashSet<Object> newcome = new HashSet<>();
  
  public int bossNum = 0;

  // Add keylistener to the game in order to obtain the user manipulation, and
  // then take specific actions accordingly.
  private final KeyListener keyListener = new KeyAdapter() {
    // We use a hashset to store all currently pressed keys.
    // This is necessary to enable multi-key listening event.
    private final Set<Integer> pressed = new HashSet<>();

    @Override
    public synchronized void keyPressed(KeyEvent e) {
      pressed.add(e.getKeyCode());
      if (!pressed.isEmpty()) {
        for (Iterator<Integer> it = pressed.iterator(); it.hasNext();) {
          switch (it.next()) {
            // Press 'W' or 'UP' buttons, the vertical direction will be -1,
            // which means that your hero plane is moving up.
            case KeyEvent.VK_W:
            case KeyEvent.VK_UP:
              hero.direction.y = -1; break;
              // Press 'A' or 'LEFT' buttons, the horizontal direction will be -1,
              // which means that your hero plane is moving to the left.
            case KeyEvent.VK_A:
            case KeyEvent.VK_LEFT:
              hero.direction.x = -1; break;
              // Press 'S' or 'DOWN' buttons, the vertical direction will be 1,
              // which means that your hero plane is moving down.
            case KeyEvent.VK_S:
            case KeyEvent.VK_DOWN:
              hero.direction.y = 1; break;
              // Press 'D' or 'RIGHT' buttons, the horizontal direction will be 1,
              // which means that your hero plane is moving to the right.
            case KeyEvent.VK_D:
            case KeyEvent.VK_RIGHT:
              hero.direction.x = 1; break;
            case KeyEvent.VK_SPACE:
              hero.fireCommand = true;
          }
        }
      }
    }

    @Override
    public synchronized void keyReleased(KeyEvent e) {
      int code = e.getKeyCode();
      pressed.remove(code);
      switch (code) {
        // Release key, then reset all direction value to zero, so that we
        // can prevent the hero from unpredictable moving.
        case KeyEvent.VK_W:
        case KeyEvent.VK_S:
        case KeyEvent.VK_UP:
        case KeyEvent.VK_DOWN:
          hero.direction.y = 0; break;
        case KeyEvent.VK_A:
        case KeyEvent.VK_D:
        case KeyEvent.VK_LEFT:
        case KeyEvent.VK_RIGHT:
          hero.direction.x = 0; break;
        case KeyEvent.VK_SPACE:
          hero.fireCommand = false;
      }
    }
  };
  
  public Canvas() {
    this.objects.add(hero); 
  }

  // Getter of the key listener.
  public KeyListener getKeyListener() {
    return this.keyListener;
  }
  // Getters of canvas width and height.
  public int getWidth() { return this.width; }
  public int getHeight() { return this.height; }

  // Getter of the hero plane.
  public HeroPlane getHeroPlane() { return hero; }

  // A very simple implementation to generate a random number given a
  // bernoulli distribution. Not perfect but enough for us.
  public static boolean bernoulli(double p) {
    if (Math.random() < p) return true;
    return false;
  }

  private void generateCharacter() {
    HashMap<String, Double> p = Setting.prob;
    if (bernoulli(p.get("EnemyPlane"))) { // Generate enemies.
      double x = (Math.random() * 0.6 + 0.2) * width;
      HashMap<String, Double> prob = Setting.subProb.get("EnemyPlane");
      if (bernoulli(prob.get("EnemyBoss")) && bossNum < 1) {
        // Generate a boss ship. Note that at most one boss ship can
        // be displayed at a time.
        newcome.add(new EnemyBoss(this, x, 0));
        bossNum++;
      } else
        newcome.add(new EnemyLightPlane(this, x, 0));
    }
    if (bernoulli(p.get("Bomb"))) { // Generate bombs.
      double x = (Math.random() * 0.6 + 0.2) * width;
      newcome.add(new Bomb(this, x, 0));
    }
    if (bernoulli(p.get("Supply"))) { // Generate supplies.
      double x = (Math.random() * 0.6 + 0.2) * width;
      newcome.add(new Supply(this, x, 0));
    }
  }

  // Draw everything including bullets, planes, supplies, etc.
  @Override
  public void paint(Graphics graphics) {
    graphics.drawImage(background, 0, 0, null);
    // The toArray method is applied, instead of iterating the set
    // itself. This is an immediate solution to fix CME.
    for (Object object : objects.toArray(new Object[0])) {
      if (object instanceof Bullet) {
        ((Bullet)object).display(graphics);
      } else if (object instanceof Plane) {
        ((Plane)object).display(graphics);
      } else if (object instanceof Bomb) {
        ((Bomb)object).display(graphics);
      } else if (object instanceof Supply) {
        ((Supply)object).display(graphics);
      }
    }

    // Draw health point and game score on the screen.
    graphics.setColor(new Color(0x000000));
    graphics.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 24));
    graphics.drawString("SCORE: " + hero.getScore(), 5, 25);
    graphics.drawString("HP: " + hero.health, 5, 55);

    if (status == Setting.GAMEOVER)
      graphics.drawImage(gameover, 0, 0, null);
  }

  public void render() {
    if (status == Setting.RUNNING) {
      generateCharacter();
      for (Object object : objects) {
        if (object instanceof Bullet) {
          ((Bullet)object).move();
          ((Bullet)object).boundaryCheck();
        } else if (object instanceof Plane) {
          ((Plane)object).move();
          ((Plane)object).boundaryCheck();
          ((Plane)object).fire();
        } else if (object instanceof Bomb) {
          ((Bomb)object).move();
          ((Bomb)object).boundaryCheck();
        } else if (object instanceof Supply) {
          ((Supply)object).move();
          ((Supply)object).boundaryCheck();
        }
      }

      for (Object object : objects) {
        if (object instanceof Plane) {
          // Check whether a plane (hero/enemy) is hit by some other
          // objects like bullets, supplies, barriers, etc.
          Plane plane = (Plane)object;
          for (Object npc : objects) {
            // If a plane is hit by an npc (bullets/bombs/...), take
            // specific actions. Different planes may react differently.
            if (plane.isHit(npc)) plane.hitBy(npc);
          }
          // The plane is shot down!
          if (plane.health <= 0) plane.explode();
        }
      }
      // Push all objects that are goind to be displayed in
      // the next frame to the objects set.
      for (Object object : newcome) objects.add(object);
      newcome.clear();

      // Remove all objects that are going to be deleted.
      // Then the trash bin will be cleared and reused.
      for (Object object : trash) objects.remove(object);
      trash.clear();
    }
    repaint();
  }

}

File: python/canvas.py
import pygame
import scipy.stats

import setting
import imgloader
from plane import *
from bomb import *
from supply import *

class Canvas:
    """This class contains the global variables associated to the game.
    These variables are going to be changed at every frame.
    """
    # The background image and gameover image.
    background = imgloader.load("background.png")
    gameover = imgloader.load("gameover.png")

    def __init__(self, graphics):
        """Create a self for graphical user interface painting.
        """
        self.graphics = graphics
        self._width, self._height = graphics.get_rect().size

        # Initialize the self.hero plane (player).
        self._hero = HeroPlane(self, *setting.HERO_INIT_POS)
        self.status = setting.RUNNING

        # ALSO DO NOT touch these variables. They are used to control what 
        # the pygame will display on the screen.
        self.planes  = { "newcome": set(), "trash": set(), "current": set() }
        self.objects = { "newcome": set(), "trash": set(), "current": set() }
        self.planes["current"].add(self.hero)

        # We use this variable to control the number of boss ships. Clearly,
        # you don't want too many boss to appear at the same time... Only one
        # will be displayed at a time. That's enough.
        self.boss_num = 0
    
    @property
    def width(self): return self._width
    @property
    def height(self): return self._height
    @property
    def hero(self): return self._hero

    def generate_character(self):
        """Randomly generate characters from the top boundary of the window.
        In general, three kinds of characters will be generated with respect
        to specific probability provided in settings.py automatically.

        Note that the probability might look very small, because it is used
        by this generator at every frame.
        """
        p = setting.PROB
        if scipy.stats.bernoulli.rvs(p["EnemyPlane"]): # Generate enemies.
            x = scipy.stats.uniform.rvs(0.2, 0.6) * self._width
            prob = setting.SUB_PROB["EnemyPlane"]["EnemyBoss"]
            if scipy.stats.bernoulli.rvs(prob) and self.boss_num < 1:
                # Generate a boss ship. Note that at most one boss ship can
                # be displayed at a time.
                self.planes["newcome"].add(EnemyBoss(self, x, 0))
                self.boss_num += 1
            else:
                self.planes["newcome"].add(EnemyLightPlane(self, x, 0))
        if scipy.stats.bernoulli.rvs(p["Bomb"]): # Generate bombs.
            x = scipy.stats.uniform.rvs(0.2, 0.6) * self._width
            self.objects["newcome"].add(Bomb(self, x, 0))
        if scipy.stats.bernoulli.rvs(p["Supply"]): # Generate supplies.
            x = scipy.stats.uniform.rvs(0.2, 0.6) * self._width
            self.objects["newcome"].add(Supply(self, x, 0))

    def paint(self):
        """Paint all objects on the screen.
        """
        # Draw everything including bullets, planes, supplies, etc.
        self.graphics.blit(self.background, (0, 0))
        for object in self.planes["current"] | self.objects["current"]:
            object.display(self.graphics)

        # Draw health point and game score on the screen.
        font = pygame.font.SysFont(pygame.font.get_default_font(), 36)
        text_sc = font.render("SCORE:" + str(self.hero.score), 1, (0, 0, 0))
        text_hp = font.render("HP:" + str(self.hero.health), 1, (0, 0, 0))
        self.graphics.blit(text_sc, (5, 25))
        self.graphics.blit(text_hp, (5, 55))
        if self.status == setting.GAMEOVER:
            self.graphics.blit(self.gameover, (0, 0))
        pygame.display.update()

    def key_event(self):
        class KeyListener:
            @staticmethod
            def response():
                for event in pygame.event.get():
                    if event.type == pygame.QUIT:
                        return False # Stop rendering and exit.
                    if event.type == pygame.KEYDOWN:
                        if event.key == pygame.K_ESCAPE:
                            return False # Stop rendering and exit.
                        if event.key == pygame.K_UP:
                            self.hero.direction.y = -1
                        elif event.key == pygame.K_LEFT:
                            self.hero.direction.x = -1
                        elif event.key == pygame.K_DOWN:
                            self.hero.direction.y = 1
                        elif event.key == pygame.K_RIGHT:
                            self.hero.direction.x = 1
                        elif event.key == pygame.K_SPACE:
                            self.hero.fire_command = True
                    if event.type == pygame.KEYUP:
                        if event.key == pygame.K_UP:
                            if self.hero.direction.y == -1:
                                self.hero.direction.y = 0
                        if event.key == pygame.K_LEFT:
                            if self.hero.direction.x == -1:
                                self.hero.direction.x = 0
                        if event.key == pygame.K_DOWN:
                            if self.hero.direction.y == 1:
                                self.hero.direction.y = 0
                        if event.key == pygame.K_RIGHT:
                            if self.hero.direction.x == 1:
                                self.hero.direction.x = 0
                        if event.key == pygame.K_SPACE:
                            self.hero.fire_command = False
                # Continue to render the game.
                return True
        return KeyListener()

    def render(self):
        if self.status == setting.RUNNING:
            self.generate_character()
            for object in self.objects["current"]:
                object.move()
                object.boundary_check()
            for object in self.planes["current"]:
                object.move()
                object.boundary_check()
                object.fire()

            for plane in self.planes["current"]:
                # Check whether a plane (self.hero/enemy) is hit by some other
                # objects like bullets, supplies, barriers, etc.
                for obj in self.objects["current"]:
                    # If a plane is hit by an npc (bullets/bombs/...), take
                    # specific actions. Different planes may react differently.
                    if plane.is_hit(obj): plane.hit_by(obj)
                # The plane is shot down!
                if plane.health <= 0: plane.explode()

            for s in [self.planes, self.objects]:
                for obj in s["newcome"]: s["current"].add(obj)
                for obj in s["trash"]: s["current"].discard(obj)
                # Reset the newcome and trash set.
                s["newcome"], s["trash"] = set(), set()
        self.paint()
